machina/apps/forum_permission/checker.py

```diff
@@ -42,127 +44,243 @@ def get_perms(self, forum):
         if not self.user.is_anonymous and not self.user.is_active:
             return []
 
-        user_model = get_user_model()
-        user_groups_related_name = user_model.groups.field.related_query_name()
+        if forum is None:
+            forum_identifier = 'global'
+        else:
+            forum_identifier = forum.id
 
-        if forum.id not in self._forum_perms_cache:
+        if forum_identifier not in self._forum_perms_cache:
             if self.user and self.user.is_superuser:
                 # The superuser has all the permissions.
                 permcodes = list(ForumPermission.objects.values_list('codename', flat=True))
             elif self.user:
-                default_auth_forum_perms = (
-                    machina_settings.DEFAULT_AUTHENTICATED_USER_FORUM_PERMISSIONS
-                )
+                perms = self.get_perms_for_forumlist([forum], None)
+                permcodes = perms[forum]
 
-                user_kwargs_filter = (
-                    {'anonymous_user': True} if self.user.is_anonymous
-                    else {'user': self.user}
-                )
+            self._forum_perms_cache[forum_identifier] = permcodes
 
-                # Fetches the permissions of the considered user for the given forum.
-                user_perms = (
-                    UserForumPermission.objects.select_related()
-                    .filter(**user_kwargs_filter)
-                    .filter(Q(forum__isnull=True) | Q(forum=forum))
-                )
+        return self._forum_perms_cache[forum_identifier]
+
+    def get_perms_for_forumlist(self, forums, perm_codenames=None):
+        """
+            Computes and returns a dictionary of [forum] to (set of permissions) for the user,
+            taking into account precendence of permissions:
+                - forum > global
+                - user > group > all_authenticated_users
+            Expects:
+            - forums to be a list of forum objects
+            - perm_codenames to be a list of permission codes (strings) to look for or None
+        """
+        globally_granted_all_users_perms = set()
+        globally_granted_group_perms, globally_nongranted_group_perms = set(), set()
+
+        user_perms = (
+            UserForumPermission.objects.select_related()
+            .filter(Q(forum__isnull=True) | Q(forum__in=forums))
+        )
+        if perm_codenames:
+            user_perms = user_perms.filter(permission__codename__in=perm_codenames)
+
+        # Do some additional filtering on (type of) user
+        if self.user.is_anonymous:
+            user_perms = user_perms.filter(anonymous_user=True)
+            all_users_perms = None
+            group_perms = None
+        else:
+            all_users_perms = user_perms.filter(authenticated_user=True)
+            user_perms = user_perms.filter(user=self.user)
```

Comment:
Perhaps you could simply fetch both types of permissions in one query here and then do the filtering in order to avoid the extra query. Something like:

```python
perms = user_perms.filter(Q(authenticated_user=True) | Q(user=self.user))
all_user_perms = [p for p in perms if p.authenticated_user]
user_perms = [p for p in perms if p.user]
```

Reply:
Thanks for the clarification. At the moment I'm at a festival so I'll look into this next week at the earliest.

Reply:
Fixed this.